src/zope/publisher/browser.py

```diff
@@ -277,69 +277,83 @@ def _decode(self, text):
+        # provides slightly better doctest-compatibility with the old code
+        # based on cgi.FieldStorage.
+        self._environ.setdefault('QUERY_STRING', '')
+
+        if self.method in _get_or_head:
+            # PEP-3333 specifies that strings must only contain codepoints
+            # representable in ISO-8859-1.
+            kwargs = {}
+            if not PYTHON2:
+                kwargs['encoding'] = 'ISO-8859-1'
+                kwargs['errors'] = 'replace'
+            items.extend(parse_qsl(
+                self._environ['QUERY_STRING'], keep_blank_values=True,
+                **kwargs))
+        elif self.method not in _get_or_head:
+            env = self._environ.copy()
+            env['wsgi.input'] = self._body_instream
+            # cgi.FieldStorage set the default Content-Type for POST
+            # requests to a "traditional" value.
+            if env.get('REQUEST_METHOD') == 'POST':
+                env.setdefault(
+                    'CONTENT_TYPE', 'application/x-www-form-urlencoded')
+            ctype = env.get('CONTENT_TYPE')
+            # Of course this isn't email, but email.message.Message has
+            # a handy Content-Type parser.
+            msg = Message()
+            msg['Content-Type'] = ctype
+            # cgi.FieldStorage treated any multipart/* Content-Type as
+            # multipart/form-data.  This seems a bit dodgy, but for
+            # compatibility we emulate it for now.
+            if ctype is not None and msg.get_content_maintype() == 'multipart':
+                msg.set_type('multipart/form-data')
+                env['CONTENT_TYPE'] = msg['Content-Type']
+            # cgi.FieldStorage allowed any HTTP method, while
+            # multipart.parse_form_data only allows POST or PUT.  However,
+            # it's helpful to support methods such as PATCH too, and
+            # multipart doesn't actually care beyond an initial check, so
+            # just pretend everything is POST from here on.
+            env['REQUEST_METHOD'] = 'POST'
+            # Monkey-patch multipart's idea of TemporaryFile to make sure
+            # all the temporary files have a name.
+            # XXX cjwatson 2020-08-04: See discussion in
+            # https://github.com/defnull/multipart/pull/22; this probably
+            # doesn't work on Windows.  Removing it would be a (small) API
+            # break, as demonstrated by BrowserTests.testFileUploadPost.
+            _orig_temporary_file = getattr(multipart, 'TemporaryFile')
+            if _orig_temporary_file is not None:
+                multipart.TemporaryFile = tempfile.NamedTemporaryFile
+            try:
+                forms, files = multipart.parse_form_data(
+                    env, charset='ISO-8859-1', memfile_limit=0)
+            finally:
+                if _orig_temporary_file is not None:
+                    multipart.TemporaryFile = _orig_temporary_file
```

Comment:
This seems frighteningly non-thread-safe.

Thread 1: _orig_temporary_file = multipart's TemporaryFile
Thread 1: multipart.TemporaryFile = tempfile.NamedTemporaryFile
Thread 2: _orig_temporary_file = multipart.TemporaryFile = tempfile.NamedTemporaryFile
Thread 1. multipart.TemporaryFile = _orig_temporary_file = multipart's TemporaryFile
Thread 2: multipart.TemporaryFile = _orig_temporary_file = tempfile.NamedTemporaryFile

and now multipart.TemporaryFile is always and forever tempfile.NamedTemporaryFile.

Seems less bad to just monkey-patch the thing at import time, if we must monkeypatch.

(also, I'm curious: under what circumstances can multipart.TemporaryFile be None?)

Reply:
The guard against it being None was to defend against future changes to multipart that switched to `NamedTemporaryFile`, which at the time I thought might be a plausible change (though multipart's maintainer has a reasonable point about that not working on Windows).

Good point on thread-safety.  What do you think about just taking the API break and dropping all this monkey-patching?  I'm not really sure what the point of exposing the `name` attribute is; it can't be used in portable code, it doesn't convey any information about the form since it's just a randomly-generated path, and `FileUpload` exposes all the file object methods that a caller would need to read the file's contents.  And in https://github.com/defnull/multipart/pull/22 the original author of this didn't really seem to know why it was there.

Reply:
For `None` to work as a guard I think you need to up update the `getattr` call and provide None as a default value.

(Oops: I looked at the getattr with a constant attribute literal and no default value and didn't wonder why you used getattr instead of regular attribute syntax.)

Reply:
I don't know of any code that expects Zope form file fields to have a `name` attribute.  (This doesn't mean such code doesn't exist, just I'm not aware of any.)

An API break seems provisionally acceptable to me.  We can always wait and see if anybody complains, and then we'll have a use case we can take upstream to `multipart`.

Reply:
OK, bumped to 6.0.0.